### ISSUE TITLE
Loss doesn't converge with mAP close to zero when running the default configured models

### ISSUE THREAD

**Issue by FELIXFENG2019** [sming256/OpenTAD#52]
When I run the default configuration files of the feature-based model Actionformer and the end-to-end Re2TAL on dataset THUMOS-14, the loss doesn't go down or down for a few iterations. Then, it increases, making the mAP close to zero during the evaluation. I followed the instructions on the README file but couldn't find out what went wrong as I was just running the original code and didn't make any changes yet.

**Comment by sming256** [comment]
Please double-check
(1) whether the VideoMAE-pretrained model is correctly loaded. Please check [here](https://github.com/sming256/OpenTAD/tree/main/configs/adatad#prepare-the-pretrained-videomae-checkpoints).
(2) whether your data is correct. You can use the released checkpoint to do the inference only. If you can not get a similar performance, your raw video could have some problems.

**Comment by FELIXFENG2019** [comment]
Thank you so much for your prompt reply! 

The story is: I want to use the Re2TAL end-to-end model based on the slowfast101 backbone and Actionformer as the detector on my own data. It doesn't works. However, my data works well on the original release of the actionformer: https://github.com/happyharrycn/actionformer_release/, the training and testing results are good. 

To start debugging: I have tried to freeze the backbone and tune the learning rate, but these don't work for me. So I changed my data to THUMOS-14 and trained the Re2TAL again and had the same problem. So, I further went back to the feature-based actionformer and still faced the same issue. 

I have loaded the pre-trained mode from: pretrained/invslowfast101_k400_pre_k400_ft_e30.pth for end-to-end training using Re2TAL, which was provided by you. I didn't see any errors when loading this pre-trained model except for the mismatched keys of the cls heads but I guess this should be fine as the pre-trained slowfast101 was just used as the backbone to extract the spatial-temporal features. 

I will try to load the released checkpoint to do the inference as you suggested. Thanks

**Comment by sming256** [comment]
Thanks for your detailed explanation. Yes, if you have loaded the pre-trained model correctly, please try to infer with the released checkpoints. If you need more help with your own data, please email me, and I will be glad to help.

**Comment by FELIXFENG2019** [comment]
> Please double-check (1) whether the VideoMAE-pretrained model is correctly loaded. Please check [here](https://github.com/sming256/OpenTAD/tree/main/configs/adatad#prepare-the-pretrained-videomae-checkpoints). (2) whether your data is correct. You can use the released checkpoint to do the inference only. If you can not get a similar performance, your raw video could have some problems.

I tried the second suggestion to load the pre-trained model first and then run the inference only and I still got zero mAP. So it seems like the thumos-14 data I use has some problems. I just followed the instructions to download the data and feed them to the model, and I haven't made any changes to the code, so do you have any idea about what could potentially goes wrong?

**Comment by sming256** [comment]
Can you provide some inference logs or inference output?

**Comment by FELIXFENG2019** [comment]
Here is the inference output I've got:

(opentad) alexa@alexa:/media/alexa/WORKSPACE/Shijia-stage-two/OpenTAD$ torchrun     --nnodes=1     --nproc_per_node=1     --rdzv_backend=c10d     --rdzv_endpoint=localhost:0     tools/test.py     configs/actionformer/thumos_i3d.py     --checkpoint exps/thumos/actionformer_i3d/gpu1_id0/checkpoint/actionformer_thumos_i3d_epoch_34_8e5ac8fd.pth
Distributed init (rank 0/1, local rank 0)
2024-12-25 15:30:15 Test INFO: Using torch version: 1.12.1, CUDA version: 11.3
2024-12-25 15:30:15 Test INFO: validation subset: 211 videos
2024-12-25 15:30:16 Test INFO: Using DDP with total 1 GPUS...
2024-12-25 15:30:16 Test INFO: Loading checkpoint from: exps/thumos/actionformer_i3d/gpu1_id0/checkpoint/actionformer_thumos_i3d_epoch_34_8e5ac8fd.pth
2024-12-25 15:30:16 Test INFO: Checkpoint is epoch 34.
2024-12-25 15:30:16 Test INFO: Using Model EMA...
2024-12-25 15:30:16 Test INFO: Testing Starts...

100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 211/211 [00:14<00:00, 14.29it/s]
2024-12-25 15:30:43 Test INFO: Evaluation starts...
2024-12-25 15:31:02 Test INFO: Loaded annotations from validation subset.
2024-12-25 15:31:02 Test INFO: Number of ground truth instances: 3325
2024-12-25 15:31:02 Test INFO: Number of predictions: 134480
2024-12-25 15:31:02 Test INFO: Fixed threshold for tiou score: [0.3, 0.4, 0.5, 0.6, 0.7]
2024-12-25 15:31:02 Test INFO: Average-mAP: 0.00 (%)
2024-12-25 15:31:02 Test INFO: mAP at tIoU 0.30 is 0.01%
2024-12-25 15:31:02 Test INFO: mAP at tIoU 0.40 is 0.00%
2024-12-25 15:31:02 Test INFO: mAP at tIoU 0.50 is 0.00%
2024-12-25 15:31:02 Test INFO: mAP at tIoU 0.60 is 0.00%
2024-12-25 15:31:02 Test INFO: mAP at tIoU 0.70 is 0.00%
2024-12-25 15:31:02 Test INFO: Testing Over...

**Comment by sming256** [comment]
I see your torch version is only 1.12.1. I haven't tested the codebase with torch version under 2.0. This could be a reason. Can you try to follow the instructions [here](https://github.com/sming256/OpenTAD/blob/main/docs/en/install.md) to set up the environment?

**Comment by FELIXFENG2019** [comment]
Yeah, I am using an old GPU 1080Ti and the Nvidia cuda version is only 11.3 and the latest pytorch version that is compatible with cuda 11.3 is the 1.12.1. I am afraid I cannot use pytorch version over 2.0. Sorry about it. But I think there is no errors or even warnings about my pytorch setting in the inference output.

**Comment by sming256** [comment]
I just tested the checkpoint with the latest code in repo, and I can get the expected mAP. I think the environment is the issue.

**Comment by FELIXFENG2019** [comment]
I re-installed my environment and exactly followed your instructions this time, but the inference results are the same:

(opentad-new) alexa@alexa:/media/alexa/WORKSPACE/Shijia-stage-two/OpenTAD$ torchrun --nnodes=1 --nproc_per_node=1 --rdzv_backend=c10d --rdzv_endpoint=localhost:0 tools/test.py configs/actionformer/thumos_i3d.py --checkpoint exps/thumos/actionformer_i3d/gpu1_id0/checkpoint/actionformer_thumos_i3d_epoch_34_8e5ac8fd.pth
master_addr is only used for static rdzv_backend and when rdzv_endpoint is not specified.
Distributed init (rank 0/1, local rank 0)
2024-12-25 17:17:32 Test INFO: Using torch version: 2.0.1, CUDA version: 11.8
2024-12-25 17:17:32 Test INFO: validation subset: 211 videos
2024-12-25 17:17:35 Test INFO: Using DDP with total 1 GPUS...
2024-12-25 17:17:35 Test INFO: Loading checkpoint from: exps/thumos/actionformer_i3d/gpu1_id0/checkpoint/actionformer_thumos_i3d_epoch_34_8e5ac8fd.pth
2024-12-25 17:17:35 Test INFO: Checkpoint is epoch 34.
2024-12-25 17:17:35 Test INFO: Using Model EMA...
2024-12-25 17:17:35 Test INFO: Testing Starts...

100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 211/211 [00:16<00:00, 12.45it/s]
2024-12-25 17:18:07 Test INFO: Evaluation starts...
2024-12-25 17:18:25 Test INFO: Loaded annotations from validation subset.
2024-12-25 17:18:25 Test INFO: Number of ground truth instances: 3325
2024-12-25 17:18:25 Test INFO: Number of predictions: 134480
2024-12-25 17:18:25 Test INFO: Fixed threshold for tiou score: [0.3, 0.4, 0.5, 0.6, 0.7]
2024-12-25 17:18:25 Test INFO: Average-mAP: 0.00 (%)
2024-12-25 17:18:25 Test INFO: mAP at tIoU 0.30 is 0.01%
2024-12-25 17:18:25 Test INFO: mAP at tIoU 0.40 is 0.00%
2024-12-25 17:18:25 Test INFO: mAP at tIoU 0.50 is 0.00%
2024-12-25 17:18:25 Test INFO: mAP at tIoU 0.60 is 0.00%
2024-12-25 17:18:25 Test INFO: mAP at tIoU 0.70 is 0.00%
2024-12-25 17:18:26 Test INFO: Testing Over...

**Comment by FELIXFENG2019** [comment]
FInally works for me:

(opentad) shijia@shijia-HP-ZBook-Power-15-6-inch-G8-Mobile-Workstation-PC:~/OpenTAD$ torchrun     --nnodes=1     --nproc_per_node=1     --rdzv_backend=c10d     --rdzv_endpoint=localhost:0     tools/test.py     configs/actionformer/thumos_i3d.py     --checkpoint exps/thumos14/actionformer_thumos_i3d_epoch_34_8e5ac8fd.pth
master_addr is only used for static rdzv_backend and when rdzv_endpoint is not specified.
Distributed init (rank 0/1, local rank 0)
2024-12-25 22:40:17 Test INFO: Using torch version: 2.0.1, CUDA version: 11.8
2024-12-25 22:40:17 Test INFO: validation subset: 211 videos
2024-12-25 22:40:17 Test INFO: Using DDP with total 1 GPUS...
2024-12-25 22:40:17 Test INFO: Loading checkpoint from: exps/thumos14/actionformer_thumos_i3d_epoch_34_8e5ac8fd.pth
2024-12-25 22:40:17 Test INFO: Checkpoint is epoch 34.
2024-12-25 22:40:17 Test INFO: Using Model EMA...
2024-12-25 22:40:17 Test INFO: Testing Starts...

100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 211/211 [00:13<00:00, 15.17it/s]
2024-12-25 22:40:32 Test INFO: Evaluation starts...
2024-12-25 22:40:45 Test INFO: Loaded annotations from validation subset.
2024-12-25 22:40:45 Test INFO: Number of ground truth instances: 3325
2024-12-25 22:40:45 Test INFO: Number of predictions: 227447
2024-12-25 22:40:45 Test INFO: Fixed threshold for tiou score: [0.3, 0.4, 0.5, 0.6, 0.7]
2024-12-25 22:40:45 Test INFO: Average-mAP: 68.44 (%)
2024-12-25 22:40:45 Test INFO: mAP at tIoU 0.30 is 83.79%
2024-12-25 22:40:45 Test INFO: mAP at tIoU 0.40 is 80.05%
2024-12-25 22:40:45 Test INFO: mAP at tIoU 0.50 is 73.16%
2024-12-25 22:40:45 Test INFO: mAP at tIoU 0.60 is 60.46%
2024-12-25 22:40:45 Test INFO: mAP at tIoU 0.70 is 44.71%
2024-12-25 22:40:45 Test INFO: Testing Over...

Seems like you are right, the code works only for a narrow range of torch and cuda versions.

**Comment by sming256** [comment]
Good to know that. Thanks for letting us know.
